src/htmlize/patchtml.ml

```diff
@@ -20,10 +19,124 @@ let edit_html ?header ?footer ~head_childs html =
+      (Some child)
+      id)
     head_childs;
   to_string html
+
+open Omd
+(* Keeps us from processing url links *)
+let is_url url =
+  (String.length url >= 7 && String.equal (String.sub url 0 7) "http://")
+  || (String.length url >= 8 && String.equal (String.sub url 0 8) "https://")
+
+(* Keeps us from processing anchors *)
+let is_anchor link =
+  String.length link >= 1 && String.equal (String.sub link 0 1) "#"
+
+(* Establish depth of given link relative to the root
+ * Then prepend the corresponding number of .. to the basename
+ * Then append index.html to fix link
+ * *)
+let handle_link link =
+
+  (* Escape through this branch if the link points to an http(s) address,
+   * or is an anchor, or link is empty
+   * *)
+  if link.destination = "" then
+    link
+  else if is_url link.destination || is_anchor link.destination then
+    link
+
+  else
+    (* Turn the link into a list in order to process it *)
+    let li = String.split_on_char '/' link.destination in
+
+    (* Filter out the empty strings resulting from forms like:
+     * path/to/directory/ = [path;to;directory;""]
+     * They're are not counted as far a depth goes.
+     * And they do not change the semantics of the link
+     * *)
+    let llink = List.filter (fun x -> x <> "." && x <> "") li in
+
+    let dep llink =
+      let neg_depth = List.filter (fun x -> x = "..") llink in
+      List.length llink - List.length neg_depth
+    in
+
+    let new_link =
+      let initial_depth = dep llink in
+
+      let rec process link =
+        match link with
+        | [] -> link
+        (* This handles the case when the target is under multiple subdirs like example/foo/bar
+         * Since we've established the depth of the path,
+         * we know when to get to the appending of "index.html" file,
+         * hence having to preppend one last level ".." before doing so.
+         * *)
+        | _dir :: dirs as path ->
+          if initial_depth = 0 || initial_depth = dep path then
+            ".."::path
+          else
+            process dirs
+      in
+      process llink in
+
+    {
+      (* [newlink] string list variable should now hold:
+       * 1. correct amound of .. to reach the root
+       * 2. the correct basename, whether it be foo/ or foo/bar/bang
+       * 3. index.html appended to the tail to fix broken MD link
+       * *)
+      link with destination =
+                  String.concat "/" (new_link @ ["index.html"]);
+    }
+
+let construct t =
+  match t with
+  | Link (attr, attr_link) ->
+    if attr_link.destination = "" then
+      Text (attr, "")
+    else
+      Link (attr, handle_link attr_link)
+  | Image (attr, attr_link) -> Image (attr, handle_link attr_link)
+  | t -> t
+
+
+let rec handle_inline = function
+  | Concat (attr, attr_inline_list) ->
+    Concat (attr, List.map handle_inline attr_inline_list)
+  | Text (attr, s) -> Text (attr, s)
+  | Emph (_attr, _attr_inline) as emph->
+    construct emph
```

Comment:
why do you call `construct` ?

Reply:
That's my bad, forgot to remove it when checking my match in `construct`